### ISSUE TITLE
Spurious sensekeys/sense identifier

### ISSUE THREAD

**Issue by 1313ou** [globalwordnet/english-wordnet#303]
This is related to but distinct from issue [157](https://github.com/globalwordnet/english-wordnet/issues/157) and issue [171](https://github.com/globalwordnet/english-wordnet/issues/171).

The 136 sensekeys/dc:identifier below refer to NOTHING in PWN31. These foreign keys are **orphaned**, point to nothing , are broken links, "Error 404", etc. 

As such, they are **garbage** and should be removed.

The way to see it with a bash script:
```
PWN31=somewhere/WordNet-3.1/dict
sks=`egrep -oh 'dc:identifier="[^"]+"' $* | sed 's#dc:identifier="\(.*\)".*#\1#g' | sed "s#&apos;#'#g" | sort`
for sk in $sks; do
	if ! egrep -q -e "^$sk " $PWN31/index.sense; then
		echo "$sk"
	fi
done
```
The list:
```
abbreviated%5:00:00:short:00
activated%5:00:00:active:00
adolescent%5:00:00:immature:00
adverse%5:00:00:unfavorable:00
antoninus_pius%1:18:00::
approximate%5:00:00:close:00
arctocephalus_townsendi%1:05:00::
august_wilhelm_von_hofmann%1:18:00::
baron_olivier_of_brighton%1:18:00::
big%5:00:00:generous:00
brief%5:00:00:short:00
bulimia_nervosa%1:26:00::
canis_lupus_arctos%1:05:00::
carl_gustaf_mosander%1:18:00::
castilian%1:18:00::
cervus_nippon%1:05:00::
cognate%5:00:01:related:00
ddc%1:06:01::
ddi%1:06:01::
dead%5:00:00:extinct:00
dead%5:00:00:insensitive:00
deep%5:00:00:low:00
dojigger%1:06:00::
earth%1:15:01::
earth%1:17:02::
epithelioid%3:01:00::
equus_africanus%1:05:02::
existing%5:00:00:present:00
extended%5:00:00:long:00
feiffer%1:18:00::
fiery%5:00:00:hot:00
filthy%5:00:00:dirty:00
flunitrazepam%1:06:00::
foul%5:00:00:dirty:00
frigid%5:00:00:cold:00
fuchsia%1:07:00::
genus_sebastes%1:05:00::
georgia_o'keeffe%1:18:00::
gioacchino_antonio_rossini%1:18:00::
glacial%5:00:00:cold:00
graeco-roman_architecture%1:09:00::
graeco-roman_wrestling%1:04:00::
grand%5:00:00:noble:00
heated%5:00:00:hot:00
Heimlich_manoeuvre%1:04:00::
helicobacter_pylori%1:05:00::
hillary_clinton%1:18:00::
hillary_rodham_clinton%1:18:00::
hoary%5:00:00:old:00
hofmann%1:18:02::
humphry_davy%1:18:00::
ichthyostega%1:05:00::
icy%5:00:00:cold:00
infighting%1:04:00::
infighting%1:04:01::
jacques_cousteau%1:18:00::
jacques_yves_cousteau%1:18:00::
jaroslav_heyrovsky%1:18:00::
jimmy_connors%1:18:00::
john_marston%1:18:00::
juan_gris%1:18:00::
jules_feiffer%1:18:00::
kb%1:23:01::
kb%1:23:03::
lackluster%5:00:00:dull:00
lacklustre%5:00:00:dull:00
lariam%1:06:00::
limber%5:00:00:flexible:00
long-range%5:00:00:long:00
louping_ill%1:26:00::
low-level%5:00:00:low:00
lusterless%5:00:00:dull:00
lustreless%5:00:00:dull:00
make_whoopee%2:41:00::
marston%1:18:00::
martial%5:00:00:military:00
moon%1:17:03::
nasty%5:00:00:dirty:00
niihau%1:15:00::
niihau_island%1:15:00::
North_Celestial_Pole%1:15:00::
Northern_Hemisphere%1:15:00::
out_of_keeping%5:00:00:incongruous:00
overstrung%5:00:00:tense:00
pale%5:00:00:colorless:00
pallid%5:00:00:colorless:00
Pampas%1:15:00::
panentheism%1:09:00::
Parietaria_difusa%1:20:00::
paroxetine%1:06:00::
pasture_land%1:15:00::
pipistrellus_hesperus%1:05:00::
poetic_justice%1:11:00::
prinivil%1:06:00::
profound%5:00:00:deep:00
prolonged%5:00:00:long:00
propranolol%1:06:00::
quiescent%5:00:00:inactive:00
raunchy%5:00:00:dirty:00
reddish-purple%5:00:00:chromatic:00
sawed-off%5:00:00:short:00
sawn-off%5:00:00:short:00
scumbag%1:18:00::
sebastes%1:05:00::
sebastes_caurinus%1:05:00::
sebastes_marinus%1:05:00::
sebastes_miniatus%1:05:00::
sebastes_ruberrimus%1:05:00::
set%5:00:00:settled:00
shigella_dysenteriae%1:05:00::
short-range%5:00:00:short:00
simon_wiesenthal%1:18:00::
sir_humphry_davy%1:18:00::
sir_richard_steele%1:18:00::
slender%5:00:00:thin:00
sluggish%5:00:00:inactive:00
smutty%5:00:00:dirty:00
snug%5:00:00:tight:00
soaring%5:00:00:high:00
South_Celestial_Pole%1:15:00::
Southern_Hemisphere%1:15:00::
steel_oneself_for%2:30:00::
steely%5:00:00:hard:00
sultry%5:00:00:hot:00
sun%1:17:02::
taut%5:00:00:tense:00
Tel_Aviv-Yafo%1:15:00::
tenuous%5:00:00:thin:00
terracotta%1:06:00::
thickened%5:00:00:thick:00
torrid%5:00:00:hot:00
toyonaka%1:15:00::
truncated%5:00:00:short:00
up_in_the_air%5:00:00:uncertain:00
veg_out%2:29:00::
white-hot%5:00:00:hot:00

```

**Comment by jmccrae** [comment]
Not referring to PWN 3.1 is not a reason by itself to remove synsets. As you have previously reported in #157, it would be desirable to add sense keys for new senses too!

The list is useful, and seems to mix a few issues:
1/ some amount of miscalculation of senses keys in the original import. These should be fixed
2/ changes caused by fixing of previous errors, e.g., "Hillary Clinton"'s name was misspelled in PWN 3.1, we have also fixed the spelling in the sense key, but this can lead to a break with PWN 3.1. Perhaps this can be documented?
3/ introduction of upper cases in sense keys that will be fixed in PR #300

**Comment by 1313ou** [comment]
What they have in common is that they point to nothing in PWN31.

If you remove them (or fix them in such a way that there is an entry in PWN index.sense), then the *dc:identifier* is a valid coherent PWN31 foreign key: it says what the sense was like in PWN31 and is a backwards reference (the fact that it is misspelt is not an issue).

If you don't, it is like a web site with invalid hyperlinks: it may or may not work. You may like to think of the url as nicely formed, but it points to nothing.

You cannot live in ambiguity. As far as EWN is concerned there are two notions of a sensekey that materialize into two sets of data. Common belief often mixes them but it's time these misconceptions were laid to rest. 

The two notions are:
- the **foreign sensekey** (the PWN umbilical cord) and 
- the **internal sensekey** 

that each abide by the following respective sets of principles:

**The foreign sensekey:**

- The sensekey points to something outside EWN, typically like ili attributes that refer to something somewhere else. It would not occur to anyone to fabricate an ili reference that points to nothing.
- Not every sense needs to have a sensekey (1076 senses lack a dc:identifier attribute)
- The lemma part of the sk need not correspond to a lemma in EWN
- The lexid part is a reference to a suffix in PWN31 lexicographer file. It follows the definition PWN gives ("lex_id is a two digit decimal integer that, when appended onto lemma , uniquely identifies a sense within a lexicographer file")
- The reference need not be unique. For instance when a PWN synset is split into 2 EWN synsets, each may keep identical references to their lineage. Another case in point is (in the the Baroque / baroque controversy) distinct senses ewn-baroque-a-02985568-01 and ewn-Baroque-a-02985568-02 both have dc:identifier="baroque%3:01:00::
- index.sense does not collect them.
- they won't evolve over time, they are references to frozen PWN31

**The internal or inner sensekey**

- **Every** sense has a sensekey. It works both ways because the sense in the source files is retrievable from it. It's a **one-one map** with no holes in it. Sensekeys point to senses **internal** to EWN
- It is a **generated artifact** in such a way that it is **unique** for each sense. These artifacts **DO NOT exist** in the source lexicographer, they are generated whenever the database is frozen. They are not hand-tweaked: the lexicographer ignores them.
- The lemma form is **retrievable** from it. The new definition may accept case (see the baroque/Baroque controversy)
- The definition of **lexid** may evolve into the following (the 1-based index of the sense in the lexical entry if multiple senses, 0 if unique sense, 0 and 1 both refer to the first sense with the added info that it is unique if 0). That's because the old lexicographer files are out of reach now. 
- It needs to be present in index.sense which is a **complete** table of sensekey-senses pairs.

As you can see. You cannot have both in one. A single sensekey cannot refer to PWN and EWN the moment EWN adds, merges, splits senses. It is a design fault (not a bug) to believe otherwise. 

These distinct principles led to XEWN (X for extended or xformed) which handles both types of sensekeys. My 2.0 schema proposition separated them with distinct sensekey and a pwn:sensekey attributes, each living in different namespaces.

The good news is that you don't have to deal with sensekeys (the PWN lexicographers didn't have to). It saves you the trouble of 
- allocating lexids
- avoiding collisions with existing sks
- dealing with the problem of case
This will alleviate your tasks and you can dedicate efforts and energy to useful purposes.

This does not preclude building a map between inner sensekeys and foreign sensekeys (XEWN does it)

**Comment by jmccrae** [comment]
In the previous release they worked as external identifiers but they are going to be internal identifiers.

As you (among others) have asked for good WNDB releases, this means that new synsets and senses will have sense keys in the WNDB release and we should thus support them as unique internal identifiers in the XML too.

Note that EWN already has a good internal identifier system in the `id` attributes that is validated for uniqueness as part of the release.

**Comment by arademaker** [comment]
I would like to just call the attention to the tool http://github.com/own-pt/mill and the format being used in our (yet another PWN fork) https://github.com/own-pt/own-en.

**Comment by jmccrae** [comment]
I have a pull request for all these changes. I did the following

Fixed: There was an error calculating the sense key, this affects satellite adjectives and a few nouns who have both lower and upper-case forms
Changed: A new file maps EWN 2020 sense keys to PWN 3.1 sense keys. This is mostly due to changes in spelling
New sense: This is a new sense in EWN 2020. There is no mapping to PWN 3.1
Already fixed: Casing error closed in #300 

> abbreviated%5:00:00:short:00
Fixed
> activated%5:00:00:active:00
Fixed
> adolescent%5:00:00:immature:00
Fixed
> adverse%5:00:00:unfavorable:00
Fixed
> antoninus_pius%1:18:00::
Changed
> approximate%5:00:00:close:00
Fixed
> arctocephalus_townsendi%1:05:00::
Changed
> august_wilhelm_von_hofmann%1:18:00::
Changed
> baron_olivier_of_brighton%1:18:00::
Changed
> big%5:00:00:generous:00
Fixed
> brief%5:00:00:short:00
Fixed
> bulimia_nervosa%1:26:00::
Changed
> canis_lupus_arctos%1:05:00::
Changed
> carl_gustaf_mosander%1:18:00::
Changed
> castilian%1:18:00::
New sense
> cervus_nippon%1:05:00::
Changed
> cognate%5:00:01:related:00
Fixed
> ddc%1:06:01::
New sense
> ddi%1:06:01::
New sense
> dead%5:00:00:extinct:00
Fixed
> dead%5:00:00:insensitive:00
Fixed
> deep%5:00:00:low:00
Fixed
> dojigger%1:06:00::
Changed
> earth%1:15:01::
Fixed
> earth%1:17:02::
Fixed
> epithelioid%3:01:00::
New sense
> equus_africanus%1:05:02::
Changed
> existing%5:00:00:present:00
Fixed 
> extended%5:00:00:long:00
Fixed
> feiffer%1:18:00::
Changed
> fiery%5:00:00:hot:00
Fixed 
> filthy%5:00:00:dirty:00
Fixed
> flunitrazepam%1:06:00::
Changed
> foul%5:00:00:dirty:00
Fixed
> frigid%5:00:00:cold:00
Fixed
> fuchsia%1:07:00::
Fixed
> genus_sebastes%1:05:00::
Changed
> georgia_o'keeffe%1:18:00::
Changed
> gioacchino_antonio_rossini%1:18:00::
Changed
> glacial%5:00:00:cold:00
Fixed
> graeco-roman_architecture%1:09:00::
New sense
> graeco-roman_wrestling%1:04:00::
New sense
> grand%5:00:00:noble:00
Fixed
> heated%5:00:00:hot:00
Fixed
> Heimlich_manoeuvre%1:04:00::
Already fixed
> helicobacter_pylori%1:05:00::
Changed
> hillary_clinton%1:18:00::
Changed
> hillary_rodham_clinton%1:18:00::
Changed
> hoary%5:00:00:old:00
Fixed
> hofmann%1:18:02::
Changed
> humphry_davy%1:18:00::
Changed
> ichthyostega%1:05:00::
Changed
> icy%5:00:00:cold:00
Fixed
> infighting%1:04:00::
Changed
> infighting%1:04:01::
Changed
> jacques_cousteau%1:18:00::
Changed
> jacques_yves_cousteau%1:18:00::
Changed
> jaroslav_heyrovsky%1:18:00::
Changed
> jimmy_connors%1:18:00::
Changed
> john_marston%1:18:00::
Changed
> juan_gris%1:18:00::
Changed 
> jules_feiffer%1:18:00::
Changed
> kb%1:23:01::
Fixed
> kb%1:23:03::
Fixed
> lackluster%5:00:00:dull:00
Fixed
> lacklustre%5:00:00:dull:00
Fixed
> lariam%1:06:00::
Changed
> limber%5:00:00:flexible:00
Fixed
> long-range%5:00:00:long:00
Fixed
> louping_ill%1:26:00::
Changed
> low-level%5:00:00:low:00
Fixed
>lusterless%5:00:00:dull:00
Fixed
>lustreless%5:00:00:dull:00
Fixed
> make_whoopee%2:41:00::
Changed
> marston%1:18:00::
Changed
> martial%5:00:00:military:00
Fixed
> moon%1:17:03::
Fixed
> nasty%5:00:00:dirty:00
Fixed
> niihau%1:15:00::
Changed
> niihau_island%1:15:00::
Changed
> North_Celestial_Pole%1:15:00::
already fixed
> Northern_Hemisphere%1:15:00::
already fixed
> out_of_keeping%5:00:00:incongruous:00
Not in PWN 3.1? New sense?
> overstrung%5:00:00:tense:00
Fixed
> pale%5:00:00:colorless:00
Fixed
> pallid%5:00:00:colorless:00
Fixed
> Pampas%1:15:00::
already fixed
> panentheism%1:09:00::
Fixed
> Parietaria_difusa%1:20:00::
already fixed
> paroxetine%1:06:00::
Changed
> pasture_land%1:15:00::
New sense
> pipistrellus_hesperus%1:05:00::
Changed
> poetic_justice%1:11:00::
Changed
> prinivil%1:06:00::
Changed
> profound%5:00:00:deep:00
Fixed 
> prolonged%5:00:00:long:00
Fixed
> propranolol%1:06:00::
Changed
> quiescent%5:00:00:inactive:00
Fixed
> raunchy%5:00:00:dirty:00
Fixed
> reddish-purple%5:00:00:chromatic:00
Not in PWN 3.1? New sense!
> sawed-off%5:00:00:short:00
Fixed
> sawn-off%5:00:00:short:00
Fixed
> scumbag%1:18:00::
Changed
> sebastes%1:05:00::
> sebastes_caurinus%1:05:00::
> sebastes_marinus%1:05:00::
> sebastes_miniatus%1:05:00::
> sebastes_ruberrimus%1:05:00::
All five changed
> set%5:00:00:settled:00
Fixed
> shigella_dysenteriae%1:05:00::
Changed
> short-range%5:00:00:short:00
Fixed 
> simon_wiesenthal%1:18:00::
Changed
> sir_humphry_davy%1:18:00::
Changed
> sir_richard_steele%1:18:00::
Changed
> slender%5:00:00:thin:00
Fixed
> sluggish%5:00:00:inactive:00
Fixed
> smutty%5:00:00:dirty:00
Fixed
> snug%5:00:00:tight:00
Fixed
> soaring%5:00:00:high:00
Fixed
> South_Celestial_Pole%1:15:00::
already fixed
> Southern_Hemisphere%1:15:00::
already fixed
> steel_oneself_for%2:30:00::
Changed
> steely%5:00:00:hard:00
Fixed
> sultry%5:00:00:hot:00
Fixed
> sun%1:17:02::
Fixed
> taut%5:00:00:tense:00
Fixed
> Tel_Aviv-Yafo%1:15:00::
already fixed
> tenuous%5:00:00:thin:00
Fixed
terracotta%1:06:00::
> thickened%5:00:00:thick:00
Fixed
> torrid%5:00:00:hot:00
Fixed
> toyonaka%1:15:00::
Changed
> truncated%5:00:00:short:00
Fixed
> up_in_the_air%5:00:00:uncertain:00
Fixed
> veg_out%2:29:00::
Changed
> white-hot%5:00:00:hot:00
Fixed
